Ieee802154 layer adaptation

GwendalLaurent opened the pull request.

GwendalLaurent committed:
adds network supervisor and adaptation of ieee802154 layers starts

GwendalLaurent committed:
encapsulation of Mac layer and Phy layer

GwendalLaurent committed:
Adds unit tests for the MAC layer for faulty and loosy PHY

GwendalLaurent committed:
simplifies IEEE 802.15.4 Stack creation

GwendalLaurent committed:
Adds a few tests for Ack requests

sylane commented on src/ieee802154.erl:
```diff
@@ -0,0 +1,92 @@
+        MacLayer::{MacModule::module(), MacParams::map(), MacState::tuple()}
+       ) -> ok.
+create_stack(IeeeParams, MacLayer) ->
+    {MacModule, MacState, MacParams} = MacLayer,
```
Tuple elements order mismatch with the spec...

sylane replied:
What is the MacState parameter ?

sylane commented on src/ieee802154.erl:
```diff
@@ -0,0 +1,92 @@
+create_stack(IeeeParams, MacLayer) ->
+    {MacModule, MacState, MacParams} = MacLayer,
+    case MacModule of
+        mac_layer -> grisp:add_device(spi2, pmod_uwb);
```
It would be better if you would call some function in the module that would do the proper setup.

sylane commented:
There should be a behaviour for the layer, and mac_layer should implement it. In addition, that would document the API.

sylane commented:
The layer should have a function used something like `State = Mod:init(Params)` and all further call should look like `State2 = Mod:fun(State, ...)`

sylane commented on src/ieee802154.erl:
```diff
@@ -0,0 +1,92 @@
+
+%%% gen_statem callbacks
+init(#{mac_layer := MAC}) ->
+    Data = #{cache => #{tx => [], rx => []}, mac_layer => MAC},
```
Initialise the layer here.

sylane commented on src/ieee802154.erl:
```diff
@@ -0,0 +1,92 @@
+
+rx(_EventType, {rx, From}, #{mac_layer := MacLayer} = Data) ->
+    io:format("Reception ~n"),
+    case MacLayer:reception() of
```
Pass the layer state to the callback.

sylane commented on src/ieee802154.erl:
```diff
@@ -0,0 +1,92 @@
+    end.
+
+tx(_EventType, {tx,FrameControl, FrameHeader, Payload, From}, #{mac_layer := MacLayer} = Data) -> 
+    case MacLayer:send_data(FrameControl, FrameHeader, Payload) of
```
Pass the layer state to the callback.

sylane commented on src/ieee802154.erl:
```diff
@@ -0,0 +1,92 @@
+    end.
+
+terminate(_Reason, _State, _Data) ->
+    ok.
```
You should probably have a terminate callback in the layer in case it has to do some cleanup.

sylane commented:
Withe the Mac layer having a state that is passed around like this, it is not a requirement for it to be a gen_server anymore. It only need to be one if it does something active without being triggered by a function call.

sylane commented:
Could there be multiple mac layers ? Do they have to be restarted independently from the ieee802154 module ? I think I would rather have the Mac layer not be a process for now. And if it was it could probably just be linked to the ieeee802154 process ? In any case, network_sup sounds too generic a name for this.

GwendalLaurent committed:
mac_layer behaviour

- Adds the mac_layer behaviour

- Adapts the stack to the new mac layer behaviour and unit tests

- Adapts the robot code to run the new mac layer on the GRiSP

- Updates documentation

- Updates Readme

GwendalLaurent committed:
Removes process for mac_layer + changes name of behaviour

sylane commented on src/gen_mac_layer.erl:
```diff
@@ -0,0 +1,42 @@
+-spec tx(State::term(), FrameControl::#frame_control{}, MacHeader::#mac_header{}, Payload::bitstring()) -> {State::term(), ok} | {State::term(), Error::atom()}.
+tx({Mod, Sub}, FrameControl, MacHeader, Payload) ->
+    case Mod:tx(Sub, FrameControl, MacHeader, Payload) of
+        {ok, Sub2} -> {{Mod, Sub2}, ok};
```
I usually would do ` {ok, Sub2} -> {ok, {Mod, Sub2}}` but this is a choice. I am not sure why, but I usually pass the state as the first argument an return it reverse: `foo(Sub, A, B) -> {ok, B2, A2, Sub2} | {error, Reason, Sub2}` 
I have no strong arguments for it though. The only thing maybe would be to keep the `{ok, ...} | {error, ...}` pattern to be kind of consistent.

sylane commented on src/gen_mac_layer.erl:
```diff
@@ -0,0 +1,42 @@
+-spec rx(State::term()) -> {State::term(), FrameControl::#frame_control{}, MacHeader::#mac_header{}, Payload::bitstring()} | {State::term(), Error::atom()}.
+rx({Mod, Sub}) ->
+    case Mod:rx(Sub) of
+        {ok, Sub2, {FrameControl, MacHeader, Payload}} -> {{Mod, Sub2}, FrameControl, MacHeader, Payload};
```
Same here, I would keek the `{ok, ...} | {error, ...}` pattern.

GwendalLaurent committed:
terminate callback and return pattern more consistent

sylane reviewed (approved):
Looks good !
If I would have a final nitpicking comment, it would be for the state return value position in the result tuple to be consistent between ok result and error result, both last or both first, and for the stop callback to be called terminate, but it is nitpicking, it looks good.

GwendalLaurent committed:
ok/error returns consistent

GwendalLaurent commented:
>  the stop callback to be called terminate

For the stop/terminate of the gen_mac_layer I followed the same architecture as gen_server. `stop` is in the module and `terminate` is the callback
